### PR TITLE
Add `hive.metastore.glue.skip-archive` config option

### PR DESCRIPTION
**Pull request by nineinchnick**
## Description

When set to true, updating tables in Glue does not create an archived
version of the table. Glue tables are updated during INSERT operations,
so this option can be used to avoid reaching the limit of table
versions, when executing many such operations.

In the second commit, I also added it to the legacy v1 Glue Metastore. I don't think it's needed there, it might help when migrating Iceberg from v1 Glue to the latest one, at least to avoid binding issues in Guice modules.

## Additional context and related issues

This is modeled after #14336

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
(x) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
## Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by ebyhr** [comment]
/test-with-secrets sha=5d538527d13412ac48610f78f48d692a49480102

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogModule.java [hunk 1]
```diff
@@ -56,10 +56,11 @@ protected void setup(Binder binder)
         binder.bind(TrinoCatalogFactory.class).to(TrinoGlueCatalogFactory.class).in(Scopes.SINGLETON);
         newExporter(binder).export(TrinoCatalogFactory.class).withGeneratedName();
 
-        install(conditionalModule(
-                IcebergGlueCatalogConfig.class,
-                IcebergGlueCatalogConfig::isSkipArchive,
-                internalBinder -> newSetBinder(internalBinder, RequestHandler2.class, ForGlueHiveMetastore.class).addBinding().toInstance(new SkipArchiveRequestHandler())));
+        GlueHiveMetastoreConfig glueHiveConfig = buildConfigObject(GlueHiveMetastoreConfig.class);
+        IcebergGlueCatalogConfig glueIcebergConfig = buildConfigObject(IcebergGlueCatalogConfig.class);
+        if (glueIcebergConfig.isSkipArchive() || !glueHiveConfig.isSkipArchive()) {
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogModule.java:61]: Not quite clear what `!glueHiveConfig.isSkipArchive()` means.
Why do we care about hive setting here?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogSkipArchive.java [hunk 2]
```diff
@@ -82,6 +82,8 @@ protected QueryRunner createQueryRunner()
                 .setIcebergProperties(
                         ImmutableMap.<String, String>builder()
                                 .put("iceberg.catalog.type", "glue")
+                                // this is not required, since iceberg.glue.skip-archive is true by default, but verify that Guice modules load correctly
```
Comment (findinpath) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogSkipArchive.java:85]: unrelated

File: docs/src/main/sphinx/object-storage/metastores.md [hunk 3]
```diff
@@ -414,6 +414,9 @@ properties:
 * - `hive.metastore.glue.partitions-segments`
   - Number of segments for partitioned Glue tables.
   - `5`
+* - `hive.metastore.glue.skip-archive`
+  - Skip archiving an old table version when creating a new version in a commit.
+  - `false`
```
[the remark this answers is not in the document]
Comment (nineinchnick) [reply docs/src/main/sphinx/object-storage/metastores.md:419]: This is for backward compatibility, I don't want to change the existing behavior. In Iceberg, we also had it `false` by default initially.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogModule.java (hunk 1 above)
Comment (nineinchnick) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogModule.java:61]: This is to avoid binding the same Guice dependency multiple times.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogSkipArchive.java (hunk 2 above)
Comment (nineinchnick) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogSkipArchive.java:85]: It is related, is the comment unclear?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogModule.java (hunk 1 above)
Comment (findinpath) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogModule.java:61]: I don't quite follow. Both settings relate to the same thing.

Why would we check for such a setting constellation?
```
iceberg.glue.skip-archive=true
hive.metastore.glue.skip-archive=false
```

Maybe we should rather add  `hive.metastore.glue.skip-archive` and use  `@LegacyConfig("iceberg.glue.skip-archive")`

File: docs/src/main/sphinx/object-storage/metastores.md [hunk 4]
```diff
@@ -414,6 +414,9 @@ properties:
 * - `hive.metastore.glue.partitions-segments`
   - Number of segments for partitioned Glue tables.
   - `5`
+* - `hive.metastore.glue.skip-archive`
+  - Skip archiving an old table version when creating a new version in a commit.
```
[the remark this answers is not in the document]
Comment (anusudarsan) [reply docs/src/main/sphinx/object-storage/metastores.md:418]: s/user can roll back/user can rollback/

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveExecutionInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.core.interceptor.ExecutionInterceptor;
+import software.amazon.awssdk.services.glue.model.UpdateTableRequest;
+
+public class GlueHiveExecutionInterceptor
+        implements ExecutionInterceptor
+{
+    private final boolean skipArchive;
+
+    GlueHiveExecutionInterceptor(GlueHiveMetastoreConfig config)
```
Comment (anusudarsan) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveExecutionInterceptor.java:27]: just pass in the boolean

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueMetastoreModule.java [hunk 6]
```diff
@@ -100,7 +100,7 @@ public static GlueCache createGlueCache(CachingHiveMetastoreConfig config, Catal
         // Note: while we could skip CachingHiveMetastoreModule altogether on workers, we retain it so that catalog
         // configuration can remain identical for all nodes, making cluster configuration easier.
         boolean enabled = nodeManager.getCurrentNode().isCoordinator() &&
-                          (metadataCacheTtl.toMillis() > 0 || statsCacheTtl.toMillis() > 0);
+                (metadataCacheTtl.toMillis() > 0 || statsCacheTtl.toMillis() > 0);
```
Comment (anusudarsan) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueMetastoreModule.java:103]: unrelated

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueMetastoreModule.java [hunk 7]
```diff
@@ -126,6 +126,7 @@ public static GlueClient createGlueClient(GlueHiveMetastoreConfig config, OpenTe
         GlueClientBuilder glue = GlueClient.builder();
 
         glue.overrideConfiguration(builder -> builder
+                .addExecutionInterceptor(new GlueHiveExecutionInterceptor(config))
```
Comment (anusudarsan) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueMetastoreModule.java:129]: isnt it better to add the interceptors (line 130) involving telemetry earlier in the chain?

**Review by anusudarsan — APPROVED** [review]
few minor comments/question otherwise lgtm

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueMetastoreModule.java (hunk 7 above)
Comment (nineinchnick) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueMetastoreModule.java:129]: I don't know

File: docs/src/main/sphinx/object-storage/metastores.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (nineinchnick) [reply docs/src/main/sphinx/object-storage/metastores.md:418]: Applied!

**Review by wendigo — APPROVED** [review]

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveExecutionInterceptor.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    GlueHiveExecutionInterceptor(boolean isSkipArchive)
+    {
+        this.skipArchive = isSkipArchive;
+    }
+
+    @Override
+    public SdkRequest modifyRequest(Context.ModifyRequest context, ExecutionAttributes executionAttributes)
+    {
+        if (context.request() instanceof UpdateTableRequest updateTableRequest) {
```
Comment (ebyhr) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveExecutionInterceptor.java:35]: I believe handling only `UpdateTableRequest` is insufficient. For instance, `BatchCreatePartitionRequest` is used when inserting into partitioned tables.

**Review by ebyhr — CHANGES_REQUESTED** [review]
Let me do `Request changes` to avoid accidental merge by other maintainers.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveExecutionInterceptor.java (hunk 8 above)
Comment (ebyhr) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveExecutionInterceptor.java:35]: The new question is why we don't list possible operations unlike `SkipArchiveRequestHandler`.

Comment (nineinchnick) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveExecutionInterceptor.java:35]: If we do that, this would make sure don't miss any operations that would benefit from this flag. But this assumes we have test coverage for all operations. I only added a test for CREATE TABLE and INSERT, should I add more DML queries?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/v1/GlueHiveMetastoreConfig.java [hunk 9]
```diff
@@ -289,6 +290,19 @@ public GlueHiveMetastoreConfig setAssumeCanonicalPartitionKeys(boolean assumeCan
...
+        return skipArchive;
+    }
+
+    @Config("hive.metastore.glue.skip-archive")
+    @ConfigDescription("Skip archiving an old table version when creating a new version in a commit")
+    public GlueHiveMetastoreConfig setSkipArchive(boolean skipArchive)
+    {
+        this.skipArchive = skipArchive;
+        return this;
+    }
```
[the remark this answers is not in the document]
Comment (nineinchnick) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/v1/GlueHiveMetastoreConfig.java]: I added this after methods for `assumeCanonicalPartitionKeys`. The order was already not consistent, so I fixed that in a separate commit.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogModule.java (hunk 1 above)
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogModule.java:61]: @nineinchnick Let's move to `@DefunctConfig` as we talked offline. `iceberg.glue.skip-archive` is enabled by default and I suppose users don't change this property.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveExecutionInterceptor.java (hunk 8 above)
Comment (ebyhr) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveExecutionInterceptor.java:35]: Let's leave it as-is.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogModule.java (hunk 1 above)
Comment (nineinchnick) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogModule.java]: Done! PTAL

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogSkipArchive.java (hunk 2 above)
Comment (nineinchnick) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogSkipArchive.java]: Not applicable anymore after cleaning up configs

**Comment by ebyhr** [comment]
/test-with-secrets sha=3c3d1dd43e6e3cbd87550ce038b5a288b627ac6f

**Comment by nineinchnick** [comment]
All green!

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogConfig.java [hunk 10]
```diff
@@ -32,17 +30,4 @@ public IcebergGlueCatalogConfig setCacheTableMetadata(boolean cacheTableMetadata
         this.cacheTableMetadata = cacheTableMetadata;
         return this;
     }
-
-    public boolean isSkipArchive()
-    {
-        return skipArchive;
-    }
-
-    @Config("iceberg.glue.skip-archive")
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogConfig.java:41]: Remove also from the `metastores.md` if the decision is to remove.
Are we talking about a "breaking change" if we remove the property?

Comment (nineinchnick) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogConfig.java:41]: Done! It's not a breaking change, as the old property will still be recognized.

**Comment by nineinchnick** [comment]
@ebyhr can you merge this?

**Comment by ebyhr** [comment]
/test-with-secrets sha=784ed976eb38742ea38a78f8fe48e74d7294404a

**Review by ebyhr — APPROVED** [review]
